src/Microsoft.SqlTools.ResourceProvider.Core/Discovery/AzureDatabaseDiscoveryProvider.cs

```diff
@@ -0,0 +1,327 @@
+            Metadata = new ExportableMetadata(
+                ServerTypes.SqlServer,
+                Categories.Azure,
+                "Microsoft.SqlServer.ConnectionServices.Azure.AzureDatabaseDiscoveryProvider");
+        }
+
+        /// <summary>
+        /// the event to raise when a database is found
+        /// </summary>
+        public event EventHandler<DatabaseInfoEventArgs> DatabaseFound;
+
+        /// <summary>
+        /// Updates the cache for current selected subscriptions
+        /// </summary>
+        /// <returns>The new cached data</returns>
+        public async Task<ServiceResponse<DatabaseInstanceInfo>> RefreshCacheAsync(CancellationToken cancellationToken)
+        {
+            ServiceResponse<DatabaseInstanceInfo> result = new ServiceResponse<DatabaseInstanceInfo>();
+
+            if (await ClearCacheAsync())
+            {
+                result = await GetDatabaseInstancesAsync(serverName: null, cancellationToken: cancellationToken);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Clears the cache for current selected subscriptions
+        /// </summary>
+        /// <returns>True if cache refreshed successfully. Otherwise returns false</returns>
+        public async Task<bool> ClearCacheAsync()
+        {
+            bool result = false;
+
+            if (AzureResourceManager != null && AccountManager != null && AzureAccountManager != null)
+            {
+                try
+                {
+                    IEnumerable<IAzureUserAccountSubscriptionContext> subscriptions = await GetSubscriptionsAsync();
+                    _cache.ClearCache(subscriptions.Select(x => x.Subscription.SubscriptionId));
+                    result = true;
+                }
+                catch (Exception ex)
+                {
+                    TraceException(TraceEventType.Error, TraceId.AzureResource, ex, "Failed to refresh the cache");
+                    result = false;
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Returns the databases for given connection info. 
+        /// The connection info should be used to make the connection for getting databases not the account manager
+        /// </summary> 
+        //public async Task<ServiceResponse<DatabaseInstanceInfo>> GetDatabaseInstancesAsync(UIConnectionInfo uiConnectionInfo, CancellationToken cancellationToken)
+        //{
+        //    ServiceResponse<DatabaseInstanceInfo> result = null;
+        //    if (DefaultDatabaseDiscoveryProvider != null && DefaultDatabaseDiscoveryProvider != this)
+        //    {
+        //        result = await DefaultDatabaseDiscoveryProvider.GetDatabaseInstancesAsync(uiConnectionInfo, cancellationToken);
+        //    }
+        //    else
+        //    {
+        //        result = new ServiceResponse<DatabaseInstanceInfo>(); //TODO: add error that we couldn't find any default database provider
```

Comment:
dead code

Reply:
I'd like to keep this code in here, so I'll add a TODO comment with what I hope to achieve. Essentially I want to move this away from UIConnectionInfo and towards our ConnectionDetails class (which will require a bit more refactoring than I can cover in this PR). With this, we can add back in support for database resource browsing so that long term we enabled Azure browse scenarios just like in SSDT.